src/main/kotlin/com/arsiu/eduhub/bpp/NotifyOnChangeAnnotationBeanPostProcessor.kt

```diff
@@ -0,0 +1,89 @@
+package com.arsiu.eduhub.bpp
+
+import com.arsiu.eduhub.annotation.NotifyOnChange
+import com.arsiu.eduhub.model.Assignment
+import com.arsiu.eduhub.model.Chapter
+import com.arsiu.eduhub.model.Course
+import com.arsiu.eduhub.model.Lesson
+import com.arsiu.eduhub.model.enums.Role
+import com.arsiu.eduhub.service.UserService
+import com.arsiu.eduhub.service.util.mail.EmailParameters
+import com.arsiu.eduhub.service.util.mail.EmailServiceImpl
+import kotlinx.coroutines.DelicateCoroutinesApi
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import org.springframework.beans.factory.config.BeanPostProcessor
+import org.springframework.stereotype.Component
+import java.lang.reflect.InvocationHandler
+import java.lang.reflect.Proxy
+import kotlin.reflect.KClass
+import kotlin.reflect.full.findAnnotation
+import kotlin.reflect.full.memberFunctions
+
+@Component
+@Suppress("SpreadOperator")
+class NotifyOnChangeAnnotationBeanPostProcessor(
+    val userService: UserService,
+    val emailServiceImpl: EmailServiceImpl
+) : BeanPostProcessor {
+
+    val map = mutableMapOf<String, KClass<*>>()
+
+    override fun postProcessBeforeInitialization(bean: Any, beanName: String): Any? {
+        val beanClass = bean::class
+
+        for (method in beanClass.java.methods) {
+            val annotation = method.getAnnotation(NotifyOnChange::class.java)
+            if (annotation != null) {
+                map[beanName] = beanClass
+                break
+            }
+        }
+
+        return bean
+    }
+
+    @OptIn(DelicateCoroutinesApi::class)
+    override fun postProcessAfterInitialization(bean: Any, beanName: String): Any? {
+        val beanClass = map[beanName]
+
+        if (beanClass != null) {
+            return Proxy.newProxyInstance(
+                beanClass.java.classLoader,
+                beanClass.java.interfaces,
+                InvocationHandler { _, method, args ->
+                    val result = method.invoke(bean, *(args ?: emptyArray()))
+
+                    val annotation = beanClass.memberFunctions.firstOrNull {
+                        (it.name == method.name) &&
+                                (it::class.java.typeParameters.contentEquals(method::class.java.typeParameters))
+                    }?.findAnnotation<NotifyOnChange>()
+
+                    if (annotation != null) {
+                        GlobalScope.launch {
+                            val info = when (result) {
+                                is Course -> result.name
+                                is Chapter -> result.course.name
+                                is Lesson -> result.chapter.course.name
+                                is Assignment -> result.lesson.chapter.course.name
+                                else -> ""
+                            }
```

Comment:
BBP is a part of Spring infrastructure. BBP must be flexible and reusable(suppose you'd like someday extract library and add it to another project).

Moreover, when you add some new entity then you must update this BPP as well to support the new entity. 
Sounds like you have strong coupling here

```kotlin
                            val info = when (result) {
                                is Course -> result.name
                                is Chapter -> result.course.name
                                is Lesson -> result.chapter.course.name
                                is Assignment -> result.lesson.chapter.course.name
                                else -> ""
                            }
```